compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/JavaClassSnapshotter.kt

```diff
@@ -51,50 +47,49 @@ object JavaClassSnapshotter {
-        val classAbiExcludingMembers = abiClass.let { snapshotJavaElement(it, it.name, includeDebugInfoInSnapshot) }
-
-        val detailedSnapshot = JavaClassMemberLevelSnapshot(classAbiExcludingMembers, fieldsAbi, methodsAbi)
         return JavaClassSnapshot(
             classId = classFile.classInfo.classId,
-            classAbiHash = JavaClassMemberLevelSnapshotExternalizer.toByteArray(detailedSnapshot).md5(),
-            classMemberLevelSnapshot = detailedSnapshot.takeIf { granularity == CLASS_MEMBER_LEVEL },
+            classAbiHash = classAbiHash,
+            classMemberLevelSnapshot = classMemberLevelSnapshot,
             supertypes = classFile.classInfo.supertypes
         )
     }
 
-    private val gson by lazy {
-        // Use serializeSpecialFloatingPointValues() to avoid this error
-        //    "java.lang.IllegalArgumentException: NaN is not a valid double value as per JSON specification. To override this behavior, use
-        //    GsonBuilder.serializeSpecialFloatingPointValues() method."
-        // on jars such as ~/.gradle/kotlin-build-dependencies/repo/kotlin.build/ideaIC/203.8084.24/artifacts/lib/rhino-1.7.12.jar.
-        GsonBuilder().serializeSpecialFloatingPointValues().create()
+    private fun snapshotClass(classNode: ClassNode): Long {
+        val classWriter = ClassWriter(0)
+        classNode.accept(classWriter)
+        return classWriter.toByteArray().hashToLong()
+    }
+
+    private fun snapshotField(fieldNode: FieldNode): Long {
+        val classNode = emptyClass()
+        classNode.fields.add(fieldNode)
+        return snapshotClass(classNode)
     }
 
-    // Same as above but with `setPrettyPrinting()`
-    private val gsonForDebug by lazy {
-        GsonBuilder().serializeSpecialFloatingPointValues()
-            .setPrettyPrinting()
-            .create()
+    private fun snapshotMethod(methodNode: MethodNode): Long {
+        val classNode = emptyClass()
+        classNode.methods.add(methodNode)
+        return snapshotClass(classNode)
     }
 
-    private fun snapshotJavaElement(
-        javaElement: Any,
-        javaElementName: String,
-        includeDebugInfoInSnapshot: Boolean
-    ): JavaElementSnapshot {
-        return if (includeDebugInfoInSnapshot) {
-            val abiValue = gsonForDebug.toJson(javaElement)
-            val abiHash = abiValue.toByteArray().md5()
-            JavaElementSnapshotForTests(javaElementName, abiHash, abiValue)
-        } else {
-            val abiValue = gson.toJson(javaElement)
-            val abiHash = abiValue.toByteArray().md5()
-            JavaElementSnapshot(javaElementName, abiHash)
-        }
+    private fun emptyClass() = ClassNode().also {
+        // We need to provide some minimal info to the class:
+        //   - Name is required.
+        //   - Class version is required if method bodies are considered, but we have removed method bodies in this class, so it's optional.
+        //   - Other info is optional.
```

Comment:
Does ASM complain about other flags e.g if you add an abstract method to a non-abstract class? One option that we discussed today is to copy all params from the original class.

Reply:
No, it doesn't complain.

I think there are a few levels of info:
  1. Name + class version only
  2. Name + class version + superName
  3. Properties provided by `ClassVisitor#visit`: name + class version + access + signature + superName + interfaces
  4. All possible class properties: The above + attrs + innerClasses + invisibleAnnotations + invisibleTypeAnnotations + module + nestHostClass + nestMembers + ...
  
I chose 1 because it's minimal and I've run this on all 600 jars of the Kotlin repo without a problem (we also have integration tests as a safety net).
 
Do you suggest a different level? (For consistency, I probably will apply the same level to both the "ClassNode" approach here and the "ClassVisitor" approach in `IncrementalJvmCache`.)

Reply:
kk, that sounds like a good check, we can keep this as-is.